lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2781,12 +2788,409 @@ ModulePass *llvm::createDxilAllocateResourcesForLibPass() {
+      SmallVector<Value *, 4> Candidates2;
+      for (auto &Select : Candidates) {
+        CreateHandleFromHeapArgs &args = HandleToArgs[Select];
+        if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+          mergeHeapArgs(Phi, args.Index, Phi->incoming_values());
+        } else if (SelectInst *Sel = dyn_cast<SelectInst>(Select)) {
+          User::op_range range = User::op_range(Sel->getOperandList() + 1,
+                                                Sel->getOperandList() + 3);
+          mergeHeapArgs(
+              Sel, args.Index,
+              range);
+        } else {
+          DXASSERT(false, "otherwise, non-select/phi in Selects set");
+        }
+        // Reload args incase it is changed in mergeHeapArgs.
+        args = HandleToArgs[Select];
+        if (args.isResolved)
+          continue;
+        Candidates2.emplace_back(Select);
+      }
+      // Some node cannot be reached.
+      if (Candidates2.size() == Candidates.size())
+        return;
+      Candidates = Candidates2;
+    }
+  }
+
+  void UpdateSelectsForHandleSelect(hlsl::OP *hlslOP) {
+    if (HandleSelects.empty())
+      return;
+    LLVMContext &Ctx = HandleSelects[0]->getContext();
+    Type *pVoidTy = Type::getVoidTy(Ctx);
+    // NOTE: currently phi of createHandleFromHeap and createHandleFromBinding
+    // is not supported.
+    Function *createHdlFromHeap =
+        hlslOP->GetOpFunc(DXIL::OpCode::CreateHandleFromHeap, pVoidTy);
+    Value *hdlFromHeapOP = hlslOP->GetI32Const(
+        static_cast<unsigned>(DXIL::OpCode::CreateHandleFromHeap));
+    for (auto &Select : HandleSelects) {
+      if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+        unsigned numIncoming = Phi->getNumIncomingValues();
+        CreateHandleFromHeapArgs &args = HandleToArgs[Phi];
+        PHINode *newPhi = cast<PHINode>(args.Index);
+        bool bAllIndexResolved = true;
+        for (unsigned j = 0; j < numIncoming; j++) {
+          // Set incoming values to undef until next pass
+          Value *V = Phi->getIncomingValue(j);
+          auto it = HandleToArgs.find(V);
+          if (it == HandleToArgs.end()) {
+            bAllIndexResolved = false;
+            break;
+          }
+          CreateHandleFromHeapArgs &itArgs = it->second;
+          if (!itArgs.isResolved) {
+            bAllIndexResolved = false;
+            break;
+          }
+          args.merge(itArgs, Errors, Phi);
+          newPhi->setIncomingValue(j, itArgs.Index);
+        }
+        if (bAllIndexResolved) {
+          IRBuilder<> B(Phi->getParent()->getFirstNonPHI());
+          B.SetCurrentDebugLocation(Phi->getDebugLoc());
+          Value *isSampler = hlslOP->GetI1Const(args.isSampler);
+          // TODO: or args.IsNonUniform with !isUniform(Phi).
+          Value *isNonUniform = hlslOP->GetI1Const(args.isNonUniform);
+          CallInst *newCI =
+              B.CreateCall(createHdlFromHeap,
+                           {hdlFromHeapOP, newPhi, isSampler, isNonUniform});
```

Comment:
Couldn't the new call become unused if it ultimately was only used by other phi/select instructions that we later replace?  Also, the original CreateHandleFromHeap calls that become dead won't be cleaned up without subsequent DCE.  Is that just what you were expecting to happen after this?  Since DCE will only occur when optimizing, perhaps we shouldn't leave extra instructions behind when we know they should be cleaned up?